text/0000-signals-in-lwc-templates.md

```diff
@@ -0,0 +1,196 @@
+
+By adopting a Signals-based reactivity in templates, LWC can overcome these challenges:
+
+1. __Framework Agnosticism__: With Signals, LWC components can interact with any external data store that follows the protocol. This change would make LWC more agnostic and flexible, allowing for easier integration with various state management patterns and libraries.
+1. __Simplified State Management__: Signals provide a straightforward way to manage reactive data. By using signals, developers can reduce complexity, making it easier to write and maintain component logic.
+1. __Performance Optimizations__: The proposed model opens up possibilities for performance optimizations by allowing selective rehydration of the UI. Components could update the DOM directly in response to state changes without a full re-render, leading to faster updates and reduced resource consumption. This has been a goal of the LWC team for a long time, small incremental improvements has been made, but this effort could accelerate the work in this area.
+
+The motivation for this RFC is to enable LWS, and developers in general, to bring stores and signal-like objects as first class citizens into the LWC ecosystem without introducing significant overhead. To improve the ergonomics of signals in LWC Template, to align LWC with more modern JavaScript practices, simplify component development, and improve performance, making LWC a developers friendly.
+
+## Detailed Design
+
+The detailed design of the new reactivity model revolves around the concept of Signals, which are reactive primitives that allow for fine-grained control over state management and UI updates. Here's how it will look in practice:
+
+### Reactivity with Signals
+
+A Signal is an object that holds a value and allows components to react to changes to that value. It exposes a `.value` property for accessing the current value, and `.subscribe` methods for responding to changes.
+
+```javascript
+import { signal } from 'some/signals';
+
+export default class ExampleComponent extends LightningElement {
+    count = signal(0);
+
+    increment() {
+        this.count.value++;
+    }
+}
+```
+
+In the template, we can bind directly to the `.value` property:
+
+```html
+<template>
+    <button onclick={increment}>Increment</button>
+    <p>{count.value}</p>
+</template>
+```
+
+### Compiler Enhancements
+
+The LWC compiler will be updated to recognize when a Signal's `.value` is used within a template. Instead of accessing the property directly at runtime, it will generate code to subscribe to the Signal's changes during component's template initialization and unsubscribe upon component's template disposal.
+
+### Store Protocol
+
+A store must implement the following protocol to be compatible with the new reactivity model:
+
+```ts
+class Store {
+    get value(): any { /* return current value */ }
+    subscribe((newValue: any) => void): () => void { /* subscribe to changes */ }
```

Comment:
Isn't there a potential for memory leaks if we don't have an `unsubscribe` method as well?

Reply:
I don't understand the question @nolanlawson. In our case, the template will always subscribe, and store the unsubscribe ref (which is returned by subscribe), and call it when disconnecting or replacing the template. No memory leaks, just like it happens with our own reactivity model today.

Reply:
Thanks, I had not thought this through. That makes sense!